[BUG] Flaky Runtime Exception while executing Painless script during Update call

**Describe the bug**

While performing updates through stored [Painless] scripts, we error out with `java.lang.UnsupportedOperationException -- java.util.Collections$UnmodifiableMap.put(Collections.java:1457) ~[?:?]` only so often.

**To Reproduce**

<details>

<summary>Steps to reproduce the behavior:</summary>

Checkout [`e4a1841`](https://github.com/opensearch-project/OpenSearch/commit/e4a1841e23dadd3351216383fdc16bf56a0b783d) from `opensearch:main`.

```sh
# run server
$ ./gradlew run

...

# Create Index
PUT sample-idx
{
  "mappings": {
    "properties": {
      "name": {
        "type": "text"
      },
      "rank": {
        "type": "integer"
      }
    }
  }
}

{
  "acknowledged": true,
  "shards_acknowledged": true,
  "index": "sample-idx"
}

# Index a Doc
POST sample-idx/_doc/1
{
  "name": "Rohit",
  "rank": 1
}

{
  "_index": "sample-idx",
  "_id": "1",
  "_version": 1,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 2,
    "failed": 0
  },
  "_seq_no": 0,
  "_primary_term": 1
}

# Store a sample script
POST _scripts/sample_script
{
  "script": {
    "lang": "painless",
    "source": """
    params.rank = 2
    """
  }
}

{
  "acknowledged": true
}

# Perform Update through Script
POST sample-idx/_update/1
{
  "script": {
    "id": "sample_script",
    "params": {
      "name": "r1walz"
    }
  }
}

# Now, it succeeds every now-and-then ...

{
  "_index": "sample-idx",
  "_id": "1",
  "_version": 2,
  "result": "updated",
  "_shards": {
    "total": 2,
    "successful": 2,
    "failed": 0
  },
  "_seq_no": 1,
  "_primary_term": 1
}

# ... but, fails more often than not

{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "failed to execute script"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "failed to execute script",
    "caused_by": {
      "type": "script_exception",
      "reason": "runtime error",
      "script_stack": [
        "java.base/java.util.Collections$UnmodifiableMap.put(Collections.java:1457)",
        "params.rank = 2\n    ",
        "      ^---- HERE"
      ],
      "script": "sample_script",
      "lang": "painless",
      "position": {
        "offset": 11,
        "start": 5,
        "end": 25
      },
      "caused_by": {
        "type": "unsupported_operation_exception",
        "reason": null
      }
    }
  },
  "status": 400
}

# Corresponding Stacktrace

java.lang.IllegalArgumentException: failed to execute script
        at org.opensearch.action.update.UpdateHelper.executeScript(UpdateHelper.java:342) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.UpdateHelper.prepareUpdateScriptRequest(UpdateHelper.java:279) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.UpdateHelper.prepare(UpdateHelper.java:109) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.UpdateHelper.prepare(UpdateHelper.java:90) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:242) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:235) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.update.TransportUpdateAction.shardOperation(TransportUpdateAction.java:90) ~[opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.action.support.single.instance.TransportInstanceSingleOperationAction$ShardTransportHandler$1.doRun(TransportInstanceSingleOperationAction.java:321) [opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:908) [opensearch-3.0.0.jar:3.0.0]
        at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) [opensearch-3.0.0.jar:3.0.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: org.opensearch.script.ScriptException: runtime error
        at org.opensearch.painless.PainlessScript.convertToScriptException(PainlessScript.java:109) ~[?:?]
        at org.opensearch.painless.PainlessScript$Script.execute(sample_script:1) ~[?:?]
        at org.opensearch.action.update.UpdateHelper.executeScript(UpdateHelper.java:339) ~[opensearch-3.0.0.jar:3.0.0]
        ... 12 more
Caused by: java.lang.UnsupportedOperationException
        at java.util.Collections$UnmodifiableMap.put(Collections.java:1457) ~[?:?]
        at org.opensearch.painless.PainlessScript$Script.execute(sample_script:12) ~[?:?]
        at org.opensearch.action.update.UpdateHelper.executeScript(UpdateHelper.java:339) ~[opensearch-3.0.0.jar:3.0.0]
        ... 12 more

```

</details>

**Expected behavior**

Painless should be execute as per expectation without encountering any bugs. In this case though, it seems throwing exception is the desired behaviour as `params` Map is immutable once created and script seems to modify it. Either way, we should throw a better cause for `unsupported_operation_exception` instead for posting `null`.

**Plugins**

Standard.

**Screenshots**

https://github.com/opensearch-project/OpenSearch/assets/31043830/562a7fa6-3231-47b0-a3fd-cb70e78d650c

**Host/Environment (please complete the following information):**

- EC2 m5.2xlarge latest default ubuntu linux AMI

**Additional context**

N/A

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -47,6 +47,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Allow clearing `remote_store.compatibility_mode` setting ([#13646](https://github.com/opensearch-project/OpenSearch/pull/13646))
 - Fix ReplicaShardBatchAllocator to batch shards without duplicates ([#13710](https://github.com/opensearch-project/OpenSearch/pull/13710))
 - Don't return negative scores from `multi_match` query with `cross_fields` type  ([#13829](https://github.com/opensearch-project/OpenSearch/pull/13829))
+- Painless: ensure type "UnmodifiableMap" for params ([#13885](https://github.com/opensearch-project/OpenSearch/pull/13885))
 - Pass parent filter to inner hit query ([#13903](https://github.com/opensearch-project/OpenSearch/pull/13903))
 - Fix NPE on restore searchable snapshot ([#13911](https://github.com/opensearch-project/OpenSearch/pull/13911))
 

--- a/modules/lang-painless/src/test/java/org/opensearch/painless/WhenThingsGoWrongTests.java
+++ b/modules/lang-painless/src/test/java/org/opensearch/painless/WhenThingsGoWrongTests.java
@@ -354,6 +354,9 @@ public class WhenThingsGoWrongTests extends ScriptTestCase {
         assertEquals(iae.getMessage(), "invalid assignment: cannot assign a value to addition operation [+]");
         iae = expectScriptThrows(IllegalArgumentException.class, () -> exec("Double.x() = 1;"));
         assertEquals(iae.getMessage(), "invalid assignment: cannot assign a value to method call [x/0]");
+
+        expectScriptThrows(UnsupportedOperationException.class, () -> exec("params['modifyingParamsMap'] = 2;"));
+        expectScriptThrows(UnsupportedOperationException.class, () -> exec("params.modifyingParamsMap = 2;"));
     }
 
     public void testCannotResolveSymbol() {

--- a/modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/17_update_error.yml
+++ b/modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/17_update_error.yml
@@ -13,3 +13,50 @@
   - match: { error.root_cause.0.position.offset: 13 }
   - match: { error.root_cause.0.position.start: 0 }
   - match: { error.root_cause.0.position.end: 38 }
+
+---
+"Test modifying params map from script leads to exception":
+  - skip:
+      features: "node_selector"
+
+  - do:
+      put_script:
+        id: "except"
+        body: {"script": {"lang": "painless", "source": "params.that = 3"}}
+
+  - do:
+      indices.create:
+        index: "test"
+        body:
+          settings:
+            index:
+              number_of_shards: 1
+              number_of_replicas: 0
+          mappings:
+            properties:
+              this:
+                type: "integer"
+              that:
+                type: "integer"
+
+  - do:
+      index:
+        index: "test"
+        id: 1
+        body: {"this": 1, "that": 2}
+
+  - do:
+      catch: /unsupported_operation_exception/
+      node_selector:
+        version: "2.15.0 - "
+      update:
+        index: "test"
+        id: 1
+        body:
+          script:
+            id: "except"
+            params: {"this": 2}
+
+  - match: { error.caused_by.position.offset: 6 }
+  - match: { error.caused_by.position.start: 0 }
+  - match: { error.caused_by.position.end: 15 }

--- a/server/src/main/java/org/opensearch/script/Script.java
+++ b/server/src/main/java/org/opensearch/script/Script.java
@@ -589,7 +589,7 @@ public final class Script implements ToXContentObject, Writeable {
         @SuppressWarnings("unchecked")
         Map<String, String> options = (Map<String, String>) (Map) in.readMap();
         this.options = options;
-        this.params = in.readMap();
+        this.params = Collections.unmodifiableMap(in.readMap());
     }
 
     @Override

--- a/server/src/main/java/org/opensearch/script/UpdateScript.java
+++ b/server/src/main/java/org/opensearch/script/UpdateScript.java
@@ -32,6 +32,7 @@
 
 package org.opensearch.script;
 
+import java.util.Collections;
 import java.util.Map;
 
 /**
@@ -53,7 +54,7 @@ public abstract class UpdateScript {
     private final Map<String, Object> ctx;
 
     public UpdateScript(Map<String, Object> params, Map<String, Object> ctx) {
-        this.params = params;
+        this.params = Collections.unmodifiableMap(params);
         this.ctx = ctx;
     }
 

